src/Fhir.Anonymizer.Core/AnonymizationConfigurations/ResourceAnonymizerContext.cs

```diff
@@ -7,55 +7,56 @@ namespace Fhir.Anonymizer.Core.AnonymizerConfigurations
         {
-            RuleList = ruleList;
-            PathSet = ruleList.Select(rule => rule.Path).ToHashSet();
+            _resourceId = resourceId;
+            _pathRuleMap = pathRuleList.ToDictionary(rule => rule.Path, rule => rule);
+            _typeRuleMap = typeRuleList.ToDictionary(rule => rule.Path, rule => rule);
         }
 
         public static ResourceAnonymizerContext Create(ElementNode root, AnonymizerConfigurationManager configurationManager)
         {
-            var rules = new List<AnonymizerRule>(configurationManager.GetPathRulesByResourceType(root.InstanceType));
-
+            var pathRules = configurationManager.GetPathRulesByResourceType(root.InstanceType); 
             var typeRules = configurationManager.GetTypeRules();
-            if (typeRules != null && typeRules.Any())
-            {
-                var rulePaths = rules.Select(rule => rule.Path).ToHashSet();
-                TransformTypeRulesToPathRules(root, typeRules, rules, rulePaths, new HashSet<string>());
-            }
-
-            return new ResourceAnonymizerContext(rules);
+            return new ResourceAnonymizerContext(root.GetNodeId(), pathRules, typeRules);
         }
 
-        private static void TransformTypeRulesToPathRules(ElementNode node, Dictionary<string, string> typeRules, List<AnonymizerRule> rules, HashSet<string> rulePaths, HashSet<string> generatedTypePaths)
+        public string GetResourceId()
         {
-            if (node.IsContainedNode() || node.IsEntryNode())
-            {
-                return;
-            }
+            return _resourceId;
+        }
 
-            string path = node.GetFhirPath();
-            if (rulePaths.Contains(path))
-            {
-                return;
-            }
+        public AnonymizerRule GetNodePathRule(ElementNode node)
+        {
+            return _pathRuleMap.GetValueOrDefault(node.GetFhirPath(), null);
+        }
 
-            if (!generatedTypePaths.Contains(path) && typeRules.ContainsKey(node.InstanceType))
-            {
-                var rule = new AnonymizerRule(path, typeRules[node.InstanceType], AnonymizerRuleType.TypeRule, node.InstanceType);
-                rules.Add(rule);
-                generatedTypePaths.Add(rule.Path);
-            }
+        public AnonymizerRule GetNodeTypeRule(ElementNode node)
+        {
+            AnonymizerRule rule = null;
+            var maxPriority = int.MaxValue;
 
-            var children = node.Children().Cast<ElementNode>();
-            foreach (var child in children)
+            var currentNode = node;
+            var typePath = string.Empty;
+            do
             {
-                TransformTypeRulesToPathRules(child, typeRules, rules, rulePaths, generatedTypePaths);
-            }
+                var path = string.IsNullOrEmpty(typePath) ? currentNode.InstanceType : $"{currentNode.InstanceType}.{typePath}";
+                var ruleForPath = _typeRuleMap.GetValueOrDefault(path, null);
```

Comment:
The naming of variable path here seems a bit strange, as it is passed as a parameter into a type rule map.